ClosedXML/Excel/Ranges/Index/XLRangeIndex.cs

```diff
@@ -0,0 +1,484 @@
+
+                if (ChildQuadrants != null)
+                {
+                    foreach (var childQuadrant in ChildQuadrants)
+                    {
+                        var childRanges = childQuadrant.GetAll();
+                        foreach (var range in childRanges)
+                            yield return range;
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Get all ranges from the quadrant and all child quadrants (recursively) that intersect the specified address.
+            /// </summary>
+            public IEnumerable<IXLRangeBase> GetIntersectedRanges(IXLRangeAddress rangeAddress)
+            {
+                if (Ranges != null)
+                {
+                    foreach (var range in Ranges)
+                    {
+                        if (range.RangeAddress.Intersects(rangeAddress))
+                            yield return range;
+                    }
+                }
+
+                if (ChildQuadrants != null)
+                {
+                    foreach (var childQuadrant in ChildQuadrants)
+                    {
+                        if (childQuadrant.Intersects(in rangeAddress))
+                        {
+                            var childRanges = childQuadrant.GetIntersectedRanges(rangeAddress);
+                            foreach (var range in childRanges)
+                                yield return range;
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Get all ranges from the quadrant and all child quadrants (recursively) that cover the specified address.
+            /// </summary>
+            public IEnumerable<IXLRangeBase> GetIntersectedRanges(IXLAddress address)
+            {
+                if (Ranges != null)
+                {
+                    foreach (var range in Ranges)
+                    {
+                        if (range.RangeAddress.Contains(address))
+                            yield return range;
+                    }
+                }
+
+                if (ChildQuadrants != null)
+                {
+                    foreach (var childQuadrant in ChildQuadrants)
+                    {
+                        if (childQuadrant.Covers(in address))
+                        {
+                            var childRanges = childQuadrant.GetIntersectedRanges(address);
+                            foreach (var range in childRanges)
+                                yield return range;
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Remove the range from the quadrant or from child quadrants (recursively).
+            /// </summary>
+            /// <returns>True if the range was removed, false if it does not exist in the QuadTree.</returns>
+            public bool Remove(IXLRangeBase range)
+            {
+                bool res = false;
```

Comment:
Check for `range == null`?

Reply:
I wanted to define `Quadrant` class as a private nested class but made it internal to be able to cover it with tests. The idea was to keep null checks and assigning defaults in the outer class (`XLRangeIndex`) so that here we could omit them. Do you think it's lame?

Reply:
Ok, if you're sure that range won't ever be null then we can skip the test.